File: packages/domain/src/inventory/product.ts
import type { ProductId, TenantId } from '../value-objects/ids.js';
import { Money } from '../value-objects/money.js';
import type { Sku } from '../value-objects/sku.js';

export interface ProductCreateInput {
  readonly id: ProductId;
  readonly name: string;
  readonly sku: Sku;
  readonly unitPrice: Money;
  readonly tenantId: TenantId;
  readonly active?: boolean;
}

/** Sellable SKU with name, identifier, and unit price. */
export class Product {
  private constructor(
    readonly id: ProductId,
    readonly name: string,
    readonly sku: Sku,
    readonly unitPrice: Money,
    readonly tenantId: TenantId,
    private readonly _active: boolean,
  ) {}

  static create(input: ProductCreateInput): Product {
    return new Product(
      input.id,
      input.name.trim(),
      input.sku,
      input.unitPrice,
      input.tenantId,
      input.active ?? true,
    );
  }

  isActive(): boolean {
    return this._active;
  }

  deactivate(): Product {
    return new Product(this.id, this.name, this.sku, this.unitPrice, this.tenantId, false);
  }
}

File: backend/crates/domain-inventory/src/product.rs
use domain_shared::{Money, TenantId};

use crate::product_id::ProductId;
use crate::sku::Sku;
use crate::unit_of_measure::UnitOfMeasure;

pub struct ProductCreateInput {
    pub id: ProductId,
    pub name: String,
    pub sku: Sku,
    pub unit_price: Money,
    pub tenant_id: TenantId,
    pub active: bool,
    pub category: Option<String>,
    pub unit_of_measure: UnitOfMeasure,
}

/// Sellable SKU with name, identifier, and unit price.
#[derive(Debug, Clone)]
pub struct Product {
    id: ProductId,
    name: String,
    sku: Sku,
    unit_price: Money,
    tenant_id: TenantId,
    active: bool,
    category: Option<String>,
    unit_of_measure: UnitOfMeasure,
}

impl Product {
    pub fn create(input: ProductCreateInput) -> Self {
        Self {
            id: input.id,
            name: input.name.trim().to_owned(),
            sku: input.sku,
            unit_price: input.unit_price,
            tenant_id: input.tenant_id,
            active: input.active,
            category: input
                .category
                .map(|c| c.trim().to_owned())
                .filter(|c| !c.is_empty()),
            unit_of_measure: input.unit_of_measure,
        }
    }

    pub fn id(&self) -> ProductId {
        self.id
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn sku(&self) -> &Sku {
        &self.sku
    }

    pub fn unit_price(&self) -> &Money {
        &self.unit_price
    }

    pub fn tenant_id(&self) -> TenantId {
        self.tenant_id
    }

    pub fn is_active(&self) -> bool {
        self.active
    }

    pub fn category(&self) -> Option<&str> {
        self.category.as_deref()
    }

    pub fn unit_of_measure(&self) -> UnitOfMeasure {
        self.unit_of_measure
    }
}
